### ISSUE TITLE
InjectOptional is stripped in IL2CPP

### ISSUE THREAD

**Issue by marsermd** [ssannandeji/Zenject-2019#405]
For some reason, `[InjectOptional]` in a method's argument is sometimes stripped off, making this argument mandatory and causing an error that this type was not bound.

So this throws an error:
```C#
public class Foo : MonoBehaviour {
    [Inject]
    public void Init([InjectOptional]Object fooDependency)
    {
        Debug.Log("Foo Created");
    }
}
```
But this works:
```C#
public class Foo : MonoBehaviour {
    [InjectOptional]
    Object fooDependency;

    [Inject]
    public void Init()
    {
        Debug.Log("Foo Created");
    }
}
```

Not sure what exactly the conditions are. In trivial cases, everything works just fine. But I'm investigating it, trying to create a minimal example.

**Comment by marsermd** [comment]
Ok, so it seems that I have narrowed it down as much as possible. It seems to me like a Unity bug, but I don't have the time right now to dig in deeper. 
So I'll leave this issue here as a warning to those who might encounter the same problem.

```C#
public interface IFoo {}

public class Foo<T> : IFoo
{
    [Inject]
    public void Init([InjectOptional]Object commandsToReplay)
    {
    }
}
```

```C#
public class FooInstaller : Installer
{
    public override void InstallBindings()
    {
        Container.BindInterfacesAndSelfTo<Foo<Object>>().AsSingle();
    }
}
```

```C#
public class GenericInstaller : MonoInstaller
{
    public override void InstallBindings()
    {
        Container
            .Bind<IFoo>()
            .ToSelf()
            .FromSubContainerResolve()
            .ByInstaller(Type.GetType("FooInstaller"))
            .AsSingle();

        Container.Resolve<IFoo>();
    }
}
```

It works fine on Mono, but on il2cpp it gives ``ZenjectException: Unable to resolve type 'UnityEngine.Object' while building object with type 'Foo`1[Bar]'. ``

There is a lot happening here:

- All the Foo/Bar classes should be separated from the other code. To do this I create a GenericInstaller, which calls FooInstaller using reflection. This ensures that stripping will be applied to Foo/Bar classes.
- There is an interface IFoo to be able to call `Container.Resolve` while leaving `Foo` separated.
- Foo is a generic class. For some reason, it is important, although the template parameter is only used to define the `Foo<T>` type!

**Comment by marsermd** [comment]
Right after posting this I realized that I haven't checked through reflection how many custom attributes there are for the parameter in `public void Init([InjectOptional]Object commandsToReplay)`.
And it turns out that Mono thinks it's 1 and IL2CPP thinks it's 0. Looks like an IL2CPP bug, since the whole function should be preserved because it has [Inject] attribute.

BTW, adding a _link.xml_ doesn't help:
```XML
<linker>
    <assembly fullname="Assembly-CSharp" preserve="all">
    </assembly>
</linker>
```

**Comment by marsermd** [comment]
Btw, my setup is
- Unity version: 2017.3.0.f3, 
- Platform: android
- C# versions: both 3.5 and 4.5

**Comment by marsermd** [comment]
UPD: I was able to reproduce the same issue without Zenject and have filed a bug report. Now waiting for Unity to say their word:)

**Comment by marsermd** [comment]
UPD#2:
Got a reply from Unity:

> Thanks for reporting this issue. We have corrected it, and the fix should be available in the **2017.3.2p1** patch release of Unity. The problem is specifically related to parameter attributes on generic methods or methods on generic types. It is possible to work around this issue if the type `Foo<T>` in this case is a non-generic type.

**Comment by svermeulen** [comment]
Thanks for following up on this issue  :)
